example/kokkos_pi/kokkos_pi.cpp

```diff
@@ -0,0 +1,381 @@
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+// This is a simple example that uses par_for() to compute whether cell
+// centers sit within a unit sphere or not.  Adding up all the
+// cells that lie within a unit sphere gives us a way to compute pi.
+//
+// Note: The goal is to use different methods of iterating through
+// mesh blocks to see what works best for different architectures.
+// While this code could be sped up by checking the innermost and
+// outermost points of a mesh block, that would defeat the purpose of
+// this program, so please do not make that change.
+//
+// Since the mesh infrastructure is not yet usable on GPUs, we create
+// mesh blocks and chain them manually.  The cell coordinates are
+// computed based on the origin of the mesh block and given cell
+// sizes.  Once we have a canonical method of using a mesh on the GPU,
+// this code will be changed to reflect that.
+//
+// Usage: examples/kokkos_pi/kokkos-pi N_Block N_Mesh N_iter
+//          N_Block = size of each mesh block on each edge
+//           N_Mesh = Number of mesh blocks along each axis
+//           N_Iter = Number of timing iterations to run
+//         [Radius] = Optional: Radius of sphere (size of cube).
+//                    Defaults to 1.0
+//
+// The unit sphere is actually a unit octant that sits within a unit
+// square which runs from (0,0,0) to (1,1,1).  Hence, in the perfect
+// case, the sum of the interior would be pi/6. Our unit cube has
+// N_Mesh*N_Block cells that span from [0,1] which gives us a
+// dimension of 1.0/(N_Mesh*N_Block) for each side of the cell and the
+// rest can be computed accordingly.  The coordinates of each cell
+// within the block can be computed as:
+//      (x0+dx*i_grid,y0+dy*j_grid,z0+dx*k_grid).
+//
+// We plan to explore using a flat range and a MDRange within par_for
+// and using flat range and MDRange in Kokkos
+//
+
+#include <stdio.h>
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include "Kokkos_Core.hpp"
+
+#include "interface/container.hpp"
+#include "interface/metadata.hpp"
+#include "kokkos_abstraction.hpp"
+#include "mesh/mesh.hpp"
+
+using Real = double;
+
+using parthenon::Container;
+using parthenon::DevExecSpace;
+using parthenon::MeshBlock;
+using parthenon::Metadata;
+using parthenon::par_for;
+using parthenon::Real;
+
+using parthenon::ParArray4D;
+using parthenon::ParArrayND;
+
+using View2D = Kokkos::View<Real **, Kokkos::LayoutRight, Kokkos::DefaultExecutionSpace>;
+
+// The result struct contains results of different tests
+typedef struct result_t {
+  std::string name; // The name of this test
+  Real pi;          // the value of pi calculated
+  Real t;           // time taken to run test
+  int iops;         // number of integer ops
+  int fops;         // number of floating point ops
+} result_t;
```

Comment:
I went with struct because it is just a data container.  However primarily because I come from a C world. :-) 
I guess if we use it outside this file, we may change it to a class.

Reply:
It's not important, obviously. I like the C++-style struct, since I think it gets the same idea accross with slightly fewer characters. In general I like C++-style structs for containers with all public scope and classes for things that might have private scope.